src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -603,10 +666,19 @@ void CallGenerator::do_late_inline_helper() {
     C->remove_macro_node(call);
   }
 
-  bool result_not_used = (callprojs.resproj == NULL || callprojs.resproj->outcnt() == 0);
-  if (is_pure_call() && result_not_used) {
+  bool call_can_eliminate = false;
+
+  if (is_pure_call()) {
+    if (is_boxing_call() && callprojs.resproj != nullptr) {
+        replace_box_to_scalar(call, callprojs.resproj);
```

Comment:
I suggest to split it into 2 parts: 
  * `bool has_non_debug_usages(Node* n)` check which ensures there are only debug usages present
  *  `scalarize_debug_usages(...)` which iterates over all usages assuming they are debug and scalarizes them.

After that you can just keep `result_not_used` as is. After all debug usages are gone, the call becomes clearly not used. And IMO `result_not_used` communicates the intention better than `call_can_eliminate`.

Reply:
Sure, I refact these codes.